### ISSUE TITLE
Missing assembly dates for some devices

### ISSUE THREAD

**Issue by CapnKernel** [SuperHouse/register#1]
In the .sql dump, devices with serial numbers from 1204 to 1212 inclusive don't have an assembly date.

Possible fixes: 

- Tell me an assembly date for these units,
- Fix dates for these devices in the google sheet, or
- Decide to set this field to allow NULLs.

**Comment by jonoxer** [comment]
Ideally we'd have a date for every board, and that should be the case in future. Unfortunately there were some batches of boards that weren't logged at the time so I didn't put in a date. We could retrospectively add dates.

I'm undecided about what the "assembled" date actually represents, because boards can proceed through multiple stages at different dates. It's very common to have PnP done on one day, then manual PTH done days later, then testing, conformal coating, etc done at stages after that. I've been conceptualising the assembly date as the date it began, ie: when parts stock is committed to the batch. This is potentially useful for billing purposes.

One option is to have multiple "assembly event" records associated with each board, and to have the date in the actual board record being the date the record was created (insertion timestamp)

TLDR: I don't know. Maybe allow NULLs for now.

**Comment by CapnKernel** [comment]
Noted.  

Those nine boards are bounded by other boards spanning about a week.  I suggest we choose an arbitrary date within that window and go forward with it.  Unless there's info to show otherwise (in which case, grab the otherwise), the precise date doesn't matter.

Regarding the meaning of assembly date, I suggest it is the date a board is assigned a serial number.  Not so much conception, gestation or birth, but a christening.

Longer term: Have a device event table.  Events have a device fk, a timestamp and a note.  Then we can show a log of what's happened, and when.

Thoughts?

**Comment by jonoxer** [comment]
I like the idea of a device events table. That would provide a great deal of flexibility. In the meantime, using the current "assembly date" to indicate when the serial number was assigned seems fine to me.

The data that was loaded was only about half what's in the spreadsheet, and was just for schema testing. Before doing the final data load I can assign "christening" dates to the boards with missing dates.

**Comment by CapnKernel** [comment]
Device events are now supported.

On import from a dumped spreadsheet, devices with ids in the range of [1204, 1212] are given an assembly date of 14 Sep 2023, and devices with ids in the range [1327, 1334] are given an assembly date of 10 Oct, 2023.  Such devices have "GUESSED ASSEMBLY DATE" added to the device notes.  Happy to consider alternatives.
